### PR TITLE
Add types to Schema namespace

### PR DESCRIPTION
**Pull request by jwage**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | yes

#### Summary

Improving types in `Doctrine\DBAL\Schema`.

### CODE REVIEW

File: lib/Doctrine/DBAL/Event/SchemaColumnDefinitionEventArgs.php [hunk 1]
```diff
@@ -33,9 +33,9 @@ class SchemaColumnDefinitionEventArgs extends SchemaEventArgs
     private $connection;
 
     /**
-     * @param mixed[] $tableColumn
-     * @param string  $table
-     * @param string  $database
+     * @param mixed[]     $tableColumn
+     * @param string      $table
+     * @param string|null $database
```
Comment (morozov) [inline lib/Doctrine/DBAL/Event/SchemaColumnDefinitionEventArgs.php]: Is it possible to not declare it nullable? A column event outside of a database doesn't make much sense.

File: lib/Doctrine/DBAL/Platforms/SqlitePlatform.php [hunk 2]
```diff
@@ -897,7 +897,7 @@ public function getAlterTableSQL(TableDiff $diff)
         if (! $this->onSchemaAlterTable($diff, $tableSql)) {
             $dataTable = new Table('__temp__' . $table->getName());
 
-            $newTable = new Table($table->getQuotedName($this), $columns, $this->getPrimaryIndexInAlteredTable($diff), [], $this->getForeignKeysInAlteredTable($diff), $table->getOptions());
+            $newTable = new Table($table->getQuotedName($this), array_values($columns), $this->getPrimaryIndexInAlteredTable($diff), [], $this->getForeignKeysInAlteredTable($diff), $table->getOptions());
```
Comment (morozov) [inline lib/Doctrine/DBAL/Platforms/SqlitePlatform.php]: After switching from `array<int, Column>` to `array<int|string, Column>`, this is probably no longer needed.

File: lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php [hunk 3]
```diff
@@ -97,9 +96,9 @@ public function tryMethod()
     /**
      * Lists the available databases for this connection.
      *
-     * @return string[]
+     * @return array<int, string>
```
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php]: This is more of a leading question than a suggestion: do we really need to enforce the type of the key if it's not used? If enforced, it may force unnecessary changes to already correct code (I'm referring to the newly introduced `array_values()`).

File: lib/Doctrine/DBAL/Schema/SQLServerSchemaManager.php [hunk 4]
```diff
@@ -56,15 +56,15 @@ public function dropDatabase($database)
     /**
      * {@inheritdoc}
      */
-    protected function _getPortableSequenceDefinition($sequence)
+    protected function _getPortableSequenceDefinition($sequence) : Sequence
```
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/SQLServerSchemaManager.php]: The argument type doesn't match the parent definition. PhpStorm in the PHP 7.1 mode reports such violations but it doesn't in PHP 7.2 mode.

File: tests/Doctrine/Tests/DBAL/Schema/ComparatorTest.php [hunk 5]
```diff
@@ -453,7 +453,7 @@ public function testCompareSequences()
         $seq1 = new Sequence('foo', 1, 1);
         $seq2 = new Sequence('foo', 1, 2);
         $seq3 = new Sequence('foo', 2, 1);
-        $seq4 = new Sequence('foo', '1', '1');
+        $seq4 = new Sequence('foo', 1, 1);
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Schema/ComparatorTest.php]: Now, this is a duplicate of `$seq1`.

File: lib/Doctrine/DBAL/Schema/Visitor/Graphviz.php [hunk 6]
```diff
@@ -116,13 +110,9 @@ private function createNode($name, $options)
...
     /**
-     * @param string   $node1
-     * @param string   $node2
-     * @param string[] $options
-     *
-     * @return string
+     * @param array<string, string> $options
      */
-    private function createNodeRelation($node1, $node2, $options)
+    private function createNodeRelation(string $node1, string $node2, $options) : string
```
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/Visitor/Graphviz.php]: Could the type of `$options` be enforced as well?

File: lib/Doctrine/DBAL/Schema/TableDiff.php [hunk 7]
```diff
@@ -130,25 +129,16 @@ public function __construct(
...
-     */
-    public function getNewName()
+    public function getNewName() : ?Identifier
     {
-        if ($this->newName === false) {
-            return false;
-        }
-
-        return new Identifier($this->newName);
+        return $this->newName !== null ? new Identifier($this->newName) : null;
```
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/TableDiff.php]: I'd rather leave the `if` instead of the ternary for better readability and more accurate code coverage.

File: lib/Doctrine/DBAL/Schema/Sequence.php [hunk 8]
```diff
@@ -22,74 +22,44 @@ class Sequence extends AbstractAsset
...
-    /**
-     * @param int $initialValue
-     *
-     * @return \Doctrine\DBAL\Schema\Sequence
-     */
-    public function setInitialValue($initialValue)
+    public function setInitialValue(int $initialValue) : self
     {
-        $this->initialValue = (int) $initialValue ?: 1;
+        $this->initialValue = $initialValue ?: 1;
```
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/Sequence.php]: Is the default value still relevant? In the past, it might cover some NULL value which is no longer relevant.

File: lib/Doctrine/DBAL/Schema/Sequence.php [hunk 9]
```diff
@@ -22,74 +22,44 @@ class Sequence extends AbstractAsset
...
-    /**
-     * @param int $allocationSize
-     *
-     * @return \Doctrine\DBAL\Schema\Sequence
-     */
-    public function setAllocationSize($allocationSize)
+    public function setAllocationSize(int $allocationSize) : self
     {
-        $this->allocationSize = (int) $allocationSize ?: 1;
+        $this->allocationSize = $allocationSize ?: 1;
```
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/Sequence.php]: Is the default value still relevant? In the past, it might cover some NULL value which is no longer relevant.

**Review by morozov — COMMENTED** [review]
Besides the need to introduce `array_values()`, looks solid. I'll see if I can run some more consistency checks on the changes later.

File: lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php (hunk 3 above)
Comment (jwage) [reply lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php]: Hmm, good question. I think yes, because without it, if you introduce new code that uses that key, phpstan won't validate that you are using the type of the key correctly.

File: lib/Doctrine/DBAL/Event/SchemaColumnDefinitionEventArgs.php (hunk 1 above)
Comment (jwage) [reply lib/Doctrine/DBAL/Event/SchemaColumnDefinitionEventArgs.php]: I will give it a try. I didn't change any signatures so far, just documented what it is today as best as I could tell.

Comment (jwage) [reply lib/Doctrine/DBAL/Event/SchemaColumnDefinitionEventArgs.php]: It is because of this and this https://github.com/doctrine/dbal/blob/develop/lib/Doctrine/DBAL/Driver.php#L53 and https://github.com/doctrine/dbal/blob/develop/lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php#L159

Comment (morozov) [reply lib/Doctrine/DBAL/Event/SchemaColumnDefinitionEventArgs.php]: Right. Please use something like 933b850a88b1151430046fe56908389cc57f18c1. If there's no DB specified in the connection parameters and not DB is currently selected, the schema manager won't be able to fetch any assets anyways. It should be an error.

File: lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php (hunk 3 above)
Comment (morozov) [reply lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php]: But you don't know the type of the key before you introduce it. And once you do, if the signature says a method accepts `array<int|string, T>`, it will have to be able to handle both `int` and `string` keys correctly.

Comment (morozov) [reply lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php]: Another reason to not introduce key types just everywhere is that `array<K, T>` implies that the method accepts (or returns) a map (correspondence) between keys and values and the order of the key-value pairs is irrelevant (or is at least secondary). While `T[]` means that it's an ordered list of `T` and the order matters while the keys don't.

File: lib/Doctrine/DBAL/Event/SchemaColumnDefinitionEventArgs.php (hunk 1 above)
Comment (jwage) [reply lib/Doctrine/DBAL/Event/SchemaColumnDefinitionEventArgs.php]: Can we make that behavior change in another PR?

Comment (morozov) [reply lib/Doctrine/DBAL/Event/SchemaColumnDefinitionEventArgs.php]: I'd rather make it now in order to not widen the type constraint. I wouldn't object if it already was `?string`. If it's an error reported by PHPStan, it's fine to temporarily suppress the error.

File: lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php (hunk 3 above)
Comment (jwage) [reply lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php]: Hmm, I am not sure. I started doing this because @Ocramius suggested it in another pull request. Without the key type, the user can pass in anything they want. Adding the type made sense because it makes the usage explicit.

Comment (morozov) [reply lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php]: Which is fine as long as the key is not used (`T[]`). Having to put `array_values()` everywhere IMO indicates a problem.

File: lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php [hunk 10]
```diff
@@ -973,4 +978,13 @@ public function getSchemaSearchPaths() : array
 
         return $match[2];
     }
+
+    private function ensureDatabase(?string $database) : string
+    {
+        if ($database === null) {
+            throw new InvalidArgumentException('Needs a database.');
```
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php]: Later, it will need a special exception.

File: tests/Doctrine/Tests/DBAL/Functional/Driver/AbstractDriverTest.php [hunk 11]
```diff
@@ -69,6 +69,6 @@ public function testReturnsDatabaseNameWithoutDatabaseNameParameter()
 
     protected static function getDatabaseNameForConnectionWithoutDatabaseNameParameter() : ?string
     {
-        return null;
+        return '';
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/Driver/AbstractDriverTest.php]: The test makes sense for the abstract driver because this is how a driver indicates that it's not connected to a DB. The test needs to be skilled in SQLite driver because it doesn't support databases. If not skipped, it'd fail with a non-in-memory database (the driver will return the database path, not null).

File: lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php (hunk 10 above)
Comment (jwage) [reply lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php]: I will go ahead and add this now.

File: UPGRADE.md [hunk 12]
```diff
@@ -1,5 +1,12 @@
 # Upgrade to 3.0
 
+## BC BREAK: Changes in the `Doctrine\DBAL\Schema` API
+
+- Method `Doctrine\DBAL\Schema\AbstractSchemaManager::_getPortableViewDefinition()` now optionally returns null instead of false.
```
Comment (morozov) [inline UPGRADE.md]: Is there a valid case where it would return `false`? In my version, the default behavior was `throw DBALException::notSupported('Views');`.

File: lib/Doctrine/DBAL/Schema/OracleSchemaManager.php [hunk 13]
```diff
@@ -288,7 +278,7 @@ protected function _getPortableDatabaseDefinition($database)
     /**
      * {@inheritdoc}
      */
-    public function createDatabase($database = null)
+    public function createDatabase(?string $database = null) : void
```
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/OracleSchemaManager.php]: `$database` should be non-nullable same as in the parent method.

File: lib/Doctrine/DBAL/Schema/Schema.php [hunk 14]
```diff
@@ -282,23 +245,19 @@ public function getSequence($sequenceName)
...
     {
         return $this->_sequences;
     }
 
     /**
      * Creates a new namespace.
      *
-     * @param string $namespaceName The name of the namespace to create.
-     *
-     * @return \Doctrine\DBAL\Schema\Schema This schema instance.
```
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/Schema.php]: It'd make sense to keep it annotated as `@return $this`. See #2384 for the reference.

File: lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php [hunk 15]
```diff
@@ -941,27 +828,27 @@ protected function _getPortableUsersList($users)
...
+     * @return array<string, View>
      */
-    protected function _getPortableViewsList($views)
+    protected function _getPortableViewsList(array $views) : array
     {
         $list = [];
         foreach ($views as $value) {
             $view = $this->_getPortableViewDefinition($value);
 
-            if (! $view) {
```
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php]: This doesn't look like a valid case. All platforms will create a view from the definition if they support views.

File: lib/Doctrine/DBAL/Schema/OracleSchemaManager.php (hunk 13 above)
Comment (jwage) [reply lib/Doctrine/DBAL/Schema/OracleSchemaManager.php]: Do you know why it was allowed to be null for Oracle? I assumed there was a reason and I thought I remembered reading something somewhere in the code about it.

File: lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php (hunk 15 above)
Comment (jwage) [reply lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php]: I will see if I can get rid of this and remove the nullable on the method too.

File: UPGRADE.md (hunk 12 above)
Comment (jwage) [reply UPGRADE.md]: I will see if I can work that out.

File: lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php (hunk 3 above)
Comment (Ocramius) [reply lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php]: If we can guarantee `array<int, T>`, then it should be like that, as it allows downstream to use it in variadic calls. Effectively, the `int` or `string` key is something that will be a BC break downstream regardless, if we change it, so if we can guarantee it by declaring it in first place, that is a win for us.

Comment (Ocramius) [reply lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php]: To make the above comment more explicit: having `array<int|string, T>` and changing implementation in the method to return `array<int, T>` to `array<string, T>` is technically not a BC break, but in practice it is one for downstream. If we can restrict types under our control, we should do so to allow these BC breaks to become explicit (when the docblock changes)

File: UPGRADE.md (hunk 12 above)
Comment (jwage) [reply UPGRADE.md]: Good call. Fixed here https://github.com/doctrine/dbal/pull/3565/commits/02d4639428a6714ad8fa2a57869916440cffa19b

File: lib/Doctrine/DBAL/Schema/Schema.php (hunk 14 above)
Comment (jwage) [reply lib/Doctrine/DBAL/Schema/Schema.php]: Fixed in https://github.com/doctrine/dbal/pull/3565/commits/05eca9944e37463a6ff0636c8c1c89a081321402

File: lib/Doctrine/DBAL/Schema/OracleSchemaManager.php (hunk 13 above)
Comment (jwage) [reply lib/Doctrine/DBAL/Schema/OracleSchemaManager.php]: Fixed in https://github.com/doctrine/dbal/pull/3565/commits/733e1b6f792adcaf3894cba2a55c25b35686e14b

File: lib/Doctrine/DBAL/Schema/Visitor/CreateSchemaSqlCollector.php [hunk 16]
```diff
@@ -47,15 +48,15 @@ public function acceptNamespace($namespaceName)
     /**
      * {@inheritdoc}
      */
-    public function acceptTable(Table $table)
+    public function acceptTable(Table $table) : void
     {
-        $this->createTableQueries = array_merge($this->createTableQueries, $this->platform->getCreateTableSQL($table));
+        $this->createTableQueries = array_merge($this->createTableQueries, array_values($this->platform->getCreateTableSQL($table)));
```
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/Visitor/CreateSchemaSqlCollector.php]: @Ocramius my major concern about adding `array<int, T>` is the need for such `array_values()` calls here and there. Specifically, in this case, it seems to massage the `string[]` returned by `AbstractPlatform::getCreateTableSQL()` to `array<int, string>` (the type of `$this->createTableQueries`) for no real reason.

File: lib/Doctrine/DBAL/Schema/Table.php [hunk 17]
```diff
@@ -57,17 +57,16 @@ class Table extends AbstractAsset
     protected $_schemaConfig = null;
 
     /**
-     * @param string                 $tableName
-     * @param Column[]               $columns
-     * @param Index[]                $indexes
-     * @param UniqueConstraint[]     $uniqueConstraints
-     * @param ForeignKeyConstraint[] $fkConstraints
-     * @param mixed[]                $options
+     * @param array<int|string, Column>               $columns
```
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/Table.php]: @Ocramius as for the definitions like this, it's not clear for the caller what an `int` or `string` keys mean. How does the caller know which keys to use in order to pass the columns to the method, and what is the meaning of those keys? The method just iterates over the array and doesn't use the key at all.

As a rule of thumb, for a union type, you should be able to explain the difference between the types and when to use what. E.g. `function bindParam(int|string $param, mixed $value)` means that the parameter is either bound by its position (`int`) or name (`string`). In the case of `array<int|string, Column>`, `int` and `string` keys mean "nothing" and "nothing" respectively.

Comment (jwage) [reply lib/Doctrine/DBAL/Schema/Table.php]: I think these cases should probably be fixed so it is only `array<int|string, Column>`. I only documented the way it is today and I think it is that way today just by chance and inconsistency due to there not being any types before.

Comment (morozov) [reply lib/Doctrine/DBAL/Schema/Table.php]: I don't agree with `array<int, Column>` in this case either. If a caller currently passes `array<string, Column>`, this constraint will force them to use `array_values()` for no real reason (the original code works just fine) — same as it's being added in this pull request.

The bottom line is, I wouldn't like us to enforce the type of the variables (keys) which we don't use..

Comment (jwage) [reply lib/Doctrine/DBAL/Schema/Table.php]: I can see both sides and I could go either way. I would rather get this merged so I will change it back.

Comment (morozov) [reply lib/Doctrine/DBAL/Schema/Table.php]: I can accept the `array<>` approach as long as the patch doesn't contain unnecessary calls to `array_values()`. They do indicate the problem and I'd rather have it fixed than masked.

Comment (jwage) [reply lib/Doctrine/DBAL/Schema/Table.php]: Take a look at this commit https://github.com/doctrine/dbal/pull/3565/commits/125265bd436523fee013e38c7e57d29b3c275995. I backed out the ones where we don't use the key and the type causes the 
 need to use `array_values`.

File: lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php [hunk 18]
```diff
@@ -100,8 +106,15 @@ public function dropAndCreateForeignKey(ForeignKeyConstraint $foreignKey, $table
      */
     public function dropForeignKey($foreignKey, $table) : void
     {
-        $tableDiff                       = $this->getTableDiffForAlterForeignKey($table);
-        $tableDiff->removedForeignKeys[] = $foreignKey;
+        $table = $this->ensureTable($table);
```
Comment (jwage) [inline lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php]: These changes were made to fix this type https://github.com/doctrine/dbal/pull/3565/commits/125265bd436523fee013e38c7e57d29b3c275995#diff-6b78caddb762256c1425fe6c316e07dcL93

File: lib/Doctrine/DBAL/Schema/SchemaDiff.php [hunk 19]
```diff
@@ -165,6 +164,6 @@ protected function _toSql(AbstractPlatform $platform, $saveMode = false)
             $sql = array_merge($sql, $platform->getAlterTableSQL($tableDiff));
         }
 
-        return $sql;
+        return array_values($sql);
```
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/SchemaDiff.php]: Is it still needed? I'd rather mark the required `AbstractPlatform` methods as returning `array<int, string>` (even though that's a different namespace) than call `array_values()` at runtime. I want to avoid the introduction of unnecessary `array_values()` because later, after all types are settled, it will be hard to figure out which of them are no longer needed.

File: UPGRADE.md [hunk 20]
```diff
@@ -1,5 +1,13 @@
 # Upgrade to 3.0
 
+## BC BREAK: Changes in the `Doctrine\DBAL\Schema` API
+
+- Method `Doctrine\DBAL\Schema\AbstractSchemaManager::_getPortableViewDefinition()` no longer optionally returns false. It will always return a `Doctrine\DBAL\Schema\View` instance.
```
Comment (morozov) [inline UPGRADE.md]: Nitpicking: although it's a change, it's not a breaking change.

**Review by morozov — COMMENTED** [review]
The recent changes look fine. We're almost there.

File: lib/Doctrine/DBAL/Schema/SchemaDiff.php (hunk 19 above)
Comment (jwage) [reply lib/Doctrine/DBAL/Schema/SchemaDiff.php]: Should be able to kill it. I'll make another commit.

File: UPGRADE.md (hunk 20 above)
Comment (morozov) [reply UPGRADE.md]: Also, should the removed methods from the description be listed here?

Comment (jwage) [reply UPGRADE.md]: Doesn't hurt. I will add.

File: lib/Doctrine/DBAL/Schema/SchemaDiff.php (hunk 19 above)
Comment (jwage) [reply lib/Doctrine/DBAL/Schema/SchemaDiff.php]: Fixed with https://github.com/doctrine/dbal/pull/3565/commits/d382e342d37227b734b0b18183a3d3fcd8923ebc

**Review by morozov — APPROVED** [review]
Awesome :+1: Size-wise, this is about 25% of the remaining type-related changes. In one shot!

**Comment by jwage** [comment]
Platforms has a bullseye on it now!
